agent_api/src/main/java/dev/aikido/agent_api/collectors/WebRequestCollector.java

```diff
@@ -38,31 +39,47 @@ public static Res report(ContextObject newContext) {
         // Increment total hits :
         StatisticsStore.incrementHits();
 
-        // Per-route IP allowlists :
-        List<Endpoint> matchedEndpoints = matchEndpoints(newContext.getRouteMetadata(), config.getEndpoints());
-        if (!ipAllowedToAccessRoute(newContext.getRemoteAddress(), matchedEndpoints)) {
+        Res endpointAllowlistRes = checkEndpointAllowlist(newContext.getRouteMetadata(), newContext.getRemoteAddress(), config);
+        Res blockedIpsRes = checkBlockedIps(newContext.getRemoteAddress(), config);
+        Res blockedUserAgentsRes = checkBlockedUserAgents(newContext.getHeader("user-agent"), config);
+
+        // make sure to follow a certain order when giving error messages.
+        if (endpointAllowlistRes != null)
+            return endpointAllowlistRes;
+        else if (blockedIpsRes != null)
+            return blockedIpsRes;
+        else return blockedUserAgentsRes;
```

Comment:
```suggestion
        // make sure to follow a certain order when giving error messages.
        Res endpointAllowlistRes = checkEndpointAllowlist(newContext.getRouteMetadata(), newContext.getRemoteAddress(), config);
        if (endpointAllowlistRes != null) {
            return endpointAllowlistRes;
        }

        Res blockedIpsRes = checkBlockedIps(newContext.getRemoteAddress(), config);
        if (blockedIpsRes != null) {
            return blockedIpsRes;
        }

        return checkBlockedUserAgents(newContext.getHeader("user-agent"), config);
```

Reply:
`config.isBlockedUserAgent(userAgent)` will be called even if `ipBlocked.blocked()`?

Reply:
(Maybe easier if we keep it more as the original and refactor in follow-up PR?)

Reply:
that's the entire idea behind it, it was to make sure we count for all of them, but since we changed this approach because it's on edge cases only, I'll change this PR as well (haven't done that yet)